src/agent.rs

```diff
@@ -0,0 +1,574 @@
+                            if let Err(e) = res {
+                                error!("An error {e} occurred in `put_mailbox`");
+                            }
+                        }
+                    }
+                }
+            }
+            info!("RxAgent thread terminated");
+            Result::<()>::Ok(())
+        });
+        this
+    }
+
+    /// Stop the `RxAgent`.
+    pub(crate) fn stop(self: &Arc<Self>) {
+        self.running.store(false, Ordering::Release);
+    }
+
+    /// Register a (`port_id`, `queue_id`) to an `RxAgent`.
+    ///
+    /// Adds a (`port_id`, `queue_id`) pair to the set to be polled.
+    pub(crate) fn register(self: &Arc<Self>, port_id: u16, queue_id: u16) -> Result<()> {
+        let _ = self
+            .tasks
+            .lock()
+            .map_err(Error::from)?
+            .insert((port_id, queue_id));
+        Ok(())
+    }
+
+    /// Unregister a (`port_id`, `queue_id`) from an `RxAgent`.
+    ///
+    /// Removes the (`port_id`, `queue_id`) pair from the polled set.
+    pub(crate) fn unregister(self: &Arc<Self>, port_id: u16, queue_id: u16) -> Result<()> {
+        let _ = self
+            .tasks
+            .lock()
+            .map_err(Error::from)?
+            .remove(&(port_id, queue_id));
+        Ok(())
+    }
+}
+
+#[allow(unsafe_code)]
+impl TxAgent {
+    /// Start a `TxBuffer`, spawn a thread to do the sending job.
+    pub(crate) fn start() -> Arc<Self> {
+        #[allow(clippy::unwrap_used)]
+        let rt = Builder::new_multi_thread()
+            .worker_threads(1)
+            .thread_name("dpdk-tx-agent")
+            .build()
+            .unwrap();
+        let this = TxAgent {
+            rt: ManuallyDrop::new(rt),
+            tasks: Mutex::new(BTreeMap::new()),
+        };
+        Arc::new(this)
+    }
+
+    /// Register a (`port_id`, `queue_id`) to a `TxAgent`.
+    ///
+    /// Spawn a new `Task` polling the given queue. It will return an error if the caller tries to
+    /// register a queue that is already registered.
+    pub(crate) fn register(
+        self: &Arc<Self>,
+        port_id: u16,
+        queue_id: u16,
+    ) -> Result<mpsc::Sender<Mbuf>> {
+        let mut tasks = self.tasks.lock().map_err(Error::from)?;
+        if tasks.contains_key(&(port_id, queue_id)) {
+            return Err(Error::Already);
+        }
+        let (tx, mut rx) = mpsc::channel::<Mbuf>(TX_CHAN_SIZE);
+        let handle = self.rt.spawn(async move {
+            let mut txbuf = TxBuffer::new(port_id, queue_id);
+            while let Some(m) = rx.recv().await {
+                let res = txbuf.buffer(m);
+                if let Err(e) = res {
+                    // TODO buffer could be full, should notify the caller.
+                    error!("An error {e} occurred in bufferring");
+                }
+            }
+        });
+        let _prev = tasks.insert((port_id, queue_id), handle);
```

Comment:
What if we register twice with different `handle`? or the same `handle`? I recommend to handle the `_prev` properly.

Reply:
Whether the (port, queue) is already registered is checked at the beginning of this function, and `_prev` is definitely `None`. An assertion is used here to make this clear.

Reply:
Then you can get the entry of this key at the beginning and assign the value to the entry if it's None, otherwise return an Error.